add cardano wallet connector trait definition

falcucci opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a Cardano wallet connector as a new trait for KubeVela, enabling applications to communicate with Cardano wallets.
- **Documentation**
	- Added documentation for the Cardano Wallet Connector trait, including configuration and example usage.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

falcucci committed:
- addition of a new file `cardano-wallet-connector.cue` for the `kubevela-addons` module

falcucci committed:
feat: add cardano wallet connector trait

- added a new 'kubectl_manifest' 'traitdefinition_cardano_wallet_connector' resource, allowing an application to connect and create a cardano wallet.
- the newly added trait applies to several workloads including deployments.apps, statefulsets.apps, daemonsets.apps, and jobs.batch.
- there are no conflicts specified with the new trait and it is marked as poddisruptive.
- the new trait schema is defined in 'cardano-wallet-connector.cue'.

falcucci committed:
feat: implement cardano wallet connector configuration

- added a new configuration script for a cardano wallet connector.
- introduced new parameters: `network` and `configcloner`.
- defined separate environments for cardano node and config cloning.
- configured volumes and environment variables for both cardano node and config cloning.
- implemented logic in `#patchconfig` to populate volumes and environment variables based on the `configcloner` value.
- designed a patch for the specification template.
- added multiple containers such as `socat`, `cardano-wallet` with their respective configurations including volume mounts, arguments etc.
- implemented an additional logic for conditional initialization of the `node-config-cloner` container.
- the `node-config-cloner` container is initialized with a git command to clone the `cardano-configurations` from the input output hk github repository.

falcucci committed:
feat: implement cardano wallet persistent volume configuration

- add a new configuration, `cardanowalletconfigs`, for the cardano wallet with defined volumes and volumemounts.
- update the volume mounts with the additional `cardanowalletconfigs` in different sections of the spec.
- change in the command parameters to include `--database` and `/wallet-db`.
- add outputs for `persistentvolumeclaim` for the volume defined in `cardanowalletconfigs`.

falcucci committed:
feat: refactor kubevela add-on for cardano wallet connector

- add a new volume named "db" in the kubevela add-on definition for the cardano wallet connector
- change the requested storage quantity for the volumes in the kubevela add-on definition from `10gi` to `100ki`
- reference storage and class name from cardanowalletconfigs in the request for resource in the kubevela add-on definition
- add a persistent volume claim for the newly added volume "db" in the kubevela add-on definition with "readwriteonce" access mode, and the storage request referenced from cardanowalletconfigs.

falcucci committed:
chore: enhance storage capacity in cardano wallet connector

- increase the storage request from `100ki` to `10gi` in the cardano wallet connector definition

falcucci committed:
feat: implement "create-wallet" container with api call

- added a new container named "create-wallet"
- the container uses the `curlimages/curl:latest` image and has the pull policy set as "always"
- the container executes a post request with headers to accept json, content-type as json, and some provided json data
- the request url is `http://localhost:8090/v2/wallets`

falcucci committed:
chore: refine cardano-wallet-connector storage and volume settings

- decreased the storage request from `10gi` to `100ki` in the cardano-wallet-connector configuration
- added volume and volumemount configurations for `cardanowalletconfigs`

falcucci commented:
atlantis plan

falcucci committed:
Merge branch 'main' into PLT-8352

falcucci commented:
atlantis plan

falcucci commented:
atlantis plan -d infra/us-east-1/prod/eks/green/kubevela-addons

falcucci commented:
atlantis plan -d infra/us-east-1/prod/eks/green/kubevela-addons

falcucci commented:
atlantis apply

falcucci commented:
atlantis plan -d infra/us-east-1/prod/eks/green/kubevela-addons

falcucci commented:
atlantis plan -d infra/us-east-1/prod/eks/green/kubevela-addons

falcucci commented:
atlantis plan -d infra/us-east-1/prod/eks/green/kubevela-addons

falcucci commented:
atlantis apply

falcucci commented:
atlantis plan -d infra/us-east-1/dev/eks/kubevela-addons

falcucci commented:
atlantis plan -d infra/us-east-1/prod/eks/green/kubevela-addons

falcucci commented:
atlantis apply -d infra/us-east-1/prod/eks/green/kubevela-addons

falcucci commented:
atlantis plan -d infra/us-east-1/dev/eks/kubevela-addons

falcucci commented:
atlantis plan

falcucci commented:
atlantis unlock

falcucci commented:
atlantis plan

falcucci commented:
atlantis plan

falcucci committed:
Merge branch 'main' into PLT-8352

falcucci commented:
atlantis plan

falcucci commented:
atlantis plan -d infra/us-east-1/dev/eks/kubevela-addons

falcucci commented:
atlantis apply

falcucci committed:
style: refactor cardano-wallet-connector configuration

- add a closing bracket in the `cardano-wallet-connector` patch block.
- remove an unnecessary closing bracket from the end of the file configuration of `cardano-wallet-connector.cue`.

falcucci commented:
atlantis plan

falcucci commented:
atlantis apply

falcucci committed:
feat: refactor cardano node and wallet connector setup

- renamed `example.yaml` to `node-connector-example.yaml` in `doc/kubevela-definitions/cardano/` directory.
- introduced new files - `node-wallet-connector.md` and `node-wallet-example.yaml` in `doc/kubevela-definitions/cardano/` directory.
- in `node-wallet-example.yaml`, defined an `application` with name `dapps-certification` having various environment variables.
- added a new rule for `localdappscertification` role to list, get, create, watch, and delete pods and logs resources.
- linked above rule to `dapps-certification` service account via rolebinding in the namespace `dapps-certification-staging`.
- added service account `dapps-certification` in namespace `dapps-certification-staging`.
- the `application` uses the `cardano-wallet-connector`, `storage`, `k8s-update-strategy`, `scaler`, `https-route` and `service-account` traits.
- the `application` has a workflow step for deploying to `local-dapps-certification`, using the `local-dapps-certification` policy.
- port `80` is exposed and set as the service's port.
- the image pull policy is set to always and `iohk-ghcr-creds` is specified for imagepullsecrets.
- various configuration maps and environment variables have been defined for this application.
- a `

falcucci committed:
docs: refactor cardano wallet connector in kubevela

- add a new documentation for the cardano wallet connector trait in kubevela with an example of usage.
- deprecate the `initcontainers` section from `cardano-wallet-connector.cue` file.

falcucci commented:
atlantis plan

falcucci commented:
atlantis apply

falcucci committed:
refactor: cardano node's volume path configuration

- change the path in the volume of the cardano node from `#configclonerconfigs` to `#cardanonodeconfigs`

falcucci commented:
atlantis plan

falcucci commented:
atlantis apply

falcucci committed:
refactor: cardano-wallet-connector for simplicity

- remove `configcloner` parameter from `kubevela-addons/definitions/cardano-wallet-connector.cue`
- simplify the composition of `volumes` and `envs` in the `patchconfig` section by removing conditions related to `configcloner`

falcucci commented:
atlantis plan

falcucci commented:
atlantis apply

falcucci committed:
refactor: integrate network parameter in cardano-wallet name

- modify name parameter in cardano-wallet definition to append network parameter

falcucci commented:
atlantis plan

falcucci commented:
atlantis apply

falcucci committed:
docs: refactor node wallet to cardano wallet in kubevela docs

- rename file `node-wallet-connector.md` to `cardano-wallet-connector.md` in `doc/kubevela-definitions/cardano/`
- rename file `node-wallet-example.yaml` to `cardano-wallet-example.yaml` in `doc/kubevela-definitions/cardano/`

falcucci committed:
chore: simplify cardano wallet connector and remove dapps-certification

- significant reduction in environment properties for the cardano wallet connector
- removed imagepullsecrets, memory, and cpu specifications from the cardano wallet connector
- the scaling properties, https route, and service account for the cardano wallet connector have been removed
- local-dapps-certification policies and workflows have been eliminated
- the serviceaccount, role, and rolebinding definitions related to dapps-certification have been deleted

falcucci committed:
-remove duplicated socat on cardano-wallet-connector.cue

- remove duplicated socat

falcucci commented:
atlantis apply

shlevy commented on doc/kubevela-definitions/cardano/cardano-wallet-connector.md:
```diff
@@ -0,0 +1,39 @@
+The ***required*** parameter for implementing this trait is:
+ - `network`: Defines the network choice for the Cardano Wallet. The available options are "preview", "preprod" and "mainnet".
+
+This configuration effectively attaches specific volumes to each container that is part of your application. Also, a `socat` container bridges the Cardano Wallet and the Cardano Node by utilizing TCP Socket over a UNIX Socket.
```
These implementation details don't belong in the "overview", and probably not in the user docs at all. Can probably just say something like "This configuration makes a cardano-wallet instance available for your application"

shlevy commented on doc/kubevela-definitions/cardano/cardano-wallet-connector.md:
```diff
@@ -0,0 +1,39 @@
+
+## Overview
+
+The Cardano Wallet Connector trait in KubeVela serves as a pass-through between your application and the Cardano wallet, using specific application characteristics and properties.
```
What does "serves as a pass-through" mean here? What "specific application characteristics and properties"?

falcucci replied:
this was too generic.

gonna write down in a simple way.

shlevy commented on doc/kubevela-definitions/cardano/cardano-wallet-example.yaml:
This file has a lot of unrelated detail. If we want to use this as an example, why not just link to it in the dapps-cert repo? Or alternatively, we can just trim it down.

shlevy commented on infra/modules/kubevela-addons/definitions/cardano-wallet-connector.cue:
```diff
@@ -0,0 +1,171 @@
+#cardanoWalletConfigs: {
+  volumes: [{
+    name: "wallet-db"
+    persistentVolumeClaim: {
```
If we're adding a `persistentVolumeClaim`, shouldn't we also add `k8s-update-strategy = Recreate`? If so, we should automatically do that and document it.

falcucci replied:
done.

shlevy commented on infra/modules/kubevela-addons/definitions/cardano-wallet-connector.cue:
```diff
@@ -0,0 +1,171 @@
+      claimName: "wallet-db"
+    }
+  }, {
+    name: "db"
```
This has nothing to do with `cardano-wallet`, this is the state dir for the dapps-certification service.

shlevy commented on infra/modules/kubevela-addons/definitions/cardano-wallet-connector.cue:
```diff
@@ -0,0 +1,171 @@
+        emptyDir: {}
+      }]
+    },
+    {
+      name:            "create-wallet"
```
Why is this here? Leftover from testing?

falcucci replied:
ops, yes. thank you!

renebarbosafl replied:
It would be nice to have some mechanism to create a wallet automatically, ideally it shouldn't be a sidecar container because only the first execution will succeed, the other ones will fail and this container will always have `CrashLoopBackOff` state.

```
% kubectl --context scde-dev -n dapps-certification-staging logs dapps-certification-6dd5b68499-q5plq -c create-wallet; echo
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   472    0   194  100   278   8864  12703 --:--:-- --:--:-- --:--:-- 22476
{"code":"wallet_already_exists","message":"This operation would yield a wallet with the following id: 73857344a0cf884fe044abfe85660cc9a81f6366 However, I already know of a wallet with this id."}
```

shlevy replied:
There shouldn't be any wallet creation here at all, any more than `sqitch` calls should be part of the RDS component

renebarbosafl replied:
I am not sure if I can agree with that because without it we have to create that wallet manually when we are deploying this application for the first time. We don't have to do nothing manually with the RDS component, it just works. :P 

Is it ok to have such manual step? If yes, it's ok for me as well. We could just get rid of that sidecar.

shlevy replied:
With the RDS component, you have to initialize your DB schema, right? In any case, yes I think it's OK, certainly it's not OK to require a wallet seed phrase  in your config :grimacing:

renebarbosafl replied:
I don't have to. The RDS creates an empty database and all the schemas are initialized by the applications themselves. Probably we will have problems with applications that don't have that kind of schema migration/initialization tho. 😢

Ok, I agree with you! Removing my request for change. :)

shlevy commented on infra/modules/kubevela-addons/definitions/cardano-wallet-connector.cue:
```diff
@@ -0,0 +1,171 @@
+parameter: {
```
Should there be a parameter to specify the port, with a sane default?

falcucci replied:
I would let this in the dapps certification side.

IMO this isn't related to the cardano-wallet itself as the same as the `db` volume, which is going to be cutted off from here.
```yaml
properties:
  image: ghcr.io/input-output-hk/plutus-certification:PLT-7784-k8s-sched
  imagePullPolicy: Always
  ports:
    - expose: true
      port: 80
      protocol: TCP
```

shlevy replied:
No, I don't mean the port of the service, I mean the cardano wallet port.

falcucci replied:
done

falcucci committed:
docs: refine kubevela documentation for cardano wallet connector

- simplify the description of the cardano wallet connector trait in the kubevela documentation
- remove the mention of `socat` container and specific volume attachment details from the document.

falcucci committed:
refactor: simplify `cardano-wallet-example.yaml` component layout

- simplified the `dapps-certification` component in the `cardano-wallet-example.yaml` file by removing numerous environment variables; `imagepullsecrets`, `cpu` and `memory` settings have been removed
- removed `storage` and numerous other types from the components section
- removed entire `policies` and `workflow` sections
- removed service account, role, and role binding definitions at the end of the file

falcucci committed:
refactor: kubernetes configuration for database handling

- removed persistentvolumeclaim `db` from `volumes`
- removed volume `db` from `volumemounts`
- removed `create-wallet` from `patch:spec:template:spec`
- removed persistentvolumeclaim `pvc-db` from `outputs`

falcucci committed:
refactor: implement recreate strategy in wallet connector spec

- modify the `spec` of the `cardano-wallet-connector.cue` definition: add a `strategy` object with the `type` set to `recreate`, and wrap the `template: spec:` block under this new `spec`.
- make use of `retainkeys` patch strategy to maintain certain keys during the patching process.
- the update does not affect the output generation related to the persistent volume claim (pvc).

falcucci commented:
atlantis apply

falcucci committed:
feat: implement 'recreate' as default strategy in wallet connector

- define "recreate" as the default deployment strategy for the cardano wallet connector trait.

falcucci commented:
atlantis apply

falcucci committed:
chore: enhance storage capacity for cardano wallet connector

- increase the storage request for cardano wallet connector from `100ki` to `10gi`

shlevy commented on doc/kubevela-definitions/cardano/cardano-wallet-connector.md:
```diff
@@ -0,0 +1,39 @@
+The ***required*** parameter for implementing this trait is:
+ - `network`: Defines the network choice for the Cardano Wallet. The available options are "preview", "preprod" and "mainnet".
+
+The *Recreate* deployment strategy has been defined as the default strategy for this trait.
```
```suggestion
Because the `cardano-wallet` stores its state in a persistent volume, this trait sets [Recreate](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#recreate-deployment) as the default deployment strategy.
```

falcucci replied:
as we override it, i've updated to:

Because the `cardano-wallet` stores its state in a persistent volume, this trait sets [Recreate](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#recreate-deployment) as its deployment strategy.

falcucci replied:
i'll put it as resolved but feel free to add a suggestion as you wish.

shlevy commented on infra/modules/kubevela-addons/definitions/cardano-wallet-connector.cue:
```diff
@@ -0,0 +1,127 @@
+        // +patchKey=name
+        env: #PatchConfig.envs
+      },
+      {
```
What will happen if the user also sets the `cardano-node` trait?

falcucci replied:
this is tricky bc somehow seems that everything works flawlessly even using both of them together.

```yaml
traits:
  - type: cardano-wallet-connector
    properties:
      network: preview
  - type: cardano-node-connector
    properties:
      network: preview
```

Anyway, even with this lucky behavior I would prefer to isolate each one of them to avoid any future problems and minimize the troubleshooting using the `conflictsWith` option for each other.

falcucci replied:
I have already [added](https://github.com/input-output-hk/sc-dev-platform/pull/106/commits/e75e0bd7365b96affd509237189f15fbbf227447) those trait definitions to conflict to each other but I dont see any treatment conditions over `conflictsWith` in the source code. For sure this needs further investigation.

For now I'll add an advice at our docs since it doesnt breakes anything, sounds good?

shlevy replied:
Sounds good

falcucci replied:
@shlevy or since it doesnt breakes anything we could leave it as it is.

let me know what you think about it.

shlevy commented on infra/modules/kubevela-addons/definitions/cardano-wallet-connector.cue:
```diff
@@ -0,0 +1,127 @@
+patch: spec: {
+  // +patchStrategy=retainKeys
+  strategy: {
+    type: "Recreate"
```
Does this override or just set a default? If the former, we should change the docs above.

falcucci replied:
overriding, updated above the docs

shlevy reviewed (dismissed):
Looking good! Small nits.

falcucci commented:
atlantis apply

falcucci commented:
atlantis plan

falcucci commented:
atlantis apply

falcucci commented:
atlantis apply

falcucci commented:
atlantis plan

falcucci commented:
atlantis apply

falcucci commented:
atlantis plan

falcucci committed:
chore: mark conflicting traits in cardano connectors

- mark "cardano-wallet-connector" as conflicting with "cardano-node-connector" in traitdefinition
- mark "cardano-node-connector" as conflicting with "cardano-wallet-connector" in traitdefinition

falcucci committed:
docs: refine cardano wallet connector deployment strategy

- clarify the rationale behind choosing the 'recreate' deployment strategy for the cardano wallet connector trait.

falcucci commented:
atlantis plan

falcucci commented:
atlantis apply

falcucci commented:
atlantis apply

falcucci commented:
atlantis apply -d infra/us-east-1/dev/eks/kubevela-addons

falcucci commented:
atlantis plan -d infra/us-east-1/dev/eks/kubevela-addons

falcucci commented:
atlantis plan -d infra/us-east-1/dev/eks/kubevela-addons

falcucci commented:
atlantis apply -d infra/us-east-1/dev/eks/kubevela-addons

falcucci commented:
atlantis apply

falcucci committed:
docs: refine `cardano-node-connector` usage guidance

- clarified the usage of the `cardano-node-connector` trait in the disclaimer, recommending its standalone usage due to shared similarities.

renebarbosafl commented:
atlantis plan

falcucci committed:
feat: implement configurable port for wallet connector

- add a "port" parameter to the cardano-wallet-connector module configuration
- include the new "port" parameter in the command line arguments passed to the wallet connector

falcucci committed:
feat: cardano wallet connector for variable port support

- add a new required parameter `port` for the cardano wallet connector trait to define the port where the cardano wallet will listen for requests with default value as 8090.

falcucci commented:
atlantis plan

falcucci commented:
atlantis plan -d infra/us-east-1/dev/eks/kubevela-addons

falcucci commented:
atlantis unlock

falcucci commented:
atlantis plan

falcucci commented:
atlantis apply

falcucci committed:
docs: refactor and update cardano wallet connector params

- refactor the parameter definition in the cardano wallet connector documentation from a list to a table format.
- change the port parameter in the cardano wallet connector definition from required to optional.

falcucci commented:
atlantis apply

falcucci committed:
feat: refine parameters and enhance test coverage

- change `port` to be a required parameter in `cardano-wallet-connector` configuration

falcucci commented:
atlantis apply

falcucci committed:
Merge branch 'main' into PLT-8352
